src/coreComponents/physicsSolvers/multiphysics/CompositionalMultiphaseReservoirAndWells.cpp

```diff
@@ -0,0 +1,404 @@
+#include "physicsSolvers/fluidFlow/wells/WellControls.hpp"
+
+namespace geosx
+{
+
+using namespace dataRepository;
+using namespace constitutive;
+
+namespace
+{
+
+template< typename COMPOSITIONAL_RESERVOIR_SOLVER > class
+  CompositionalCatalogNames {};
+
+template<> class CompositionalCatalogNames< CompositionalMultiphaseFVM >
+{
+public:
+  static string name() { return "CompositionalMultiphaseReservoirFVM"; }
+};
+template<> class CompositionalCatalogNames< CompositionalMultiphaseHybridFVM >
+{
+public:
+  static string name() { return "CompositionalMultiphaseReservoirHybridFVM"; }
+};
+
+}
+
+// provide a definition for catalogName()
+template< typename COMPOSITIONAL_RESERVOIR_SOLVER >
+string
+CompositionalMultiphaseReservoirAndWells< COMPOSITIONAL_RESERVOIR_SOLVER >::
+catalogName()
+{
+  return CompositionalCatalogNames< COMPOSITIONAL_RESERVOIR_SOLVER >::name();
+}
+
+
+template< typename COMPOSITIONAL_RESERVOIR_SOLVER >
+CompositionalMultiphaseReservoirAndWells< COMPOSITIONAL_RESERVOIR_SOLVER >::
+CompositionalMultiphaseReservoirAndWells( const string & name,
+                                          Group * const parent )
+  : Base( name, parent )
+{}
+
+template< typename COMPOSITIONAL_RESERVOIR_SOLVER >
+CompositionalMultiphaseReservoirAndWells< COMPOSITIONAL_RESERVOIR_SOLVER >::
+~CompositionalMultiphaseReservoirAndWells()
+{}
+
+template< typename COMPOSITIONAL_RESERVOIR_SOLVER >
+void
+CompositionalMultiphaseReservoirAndWells< COMPOSITIONAL_RESERVOIR_SOLVER >::
+postProcessInput()
+{
+  Base::postProcessInput();
+
+  COMPOSITIONAL_RESERVOIR_SOLVER * reservoirSolver =
+    &this->getParent().template getGroup< COMPOSITIONAL_RESERVOIR_SOLVER >( m_reservoirSolverName );
+  std::get< toUnderlying< typename Base::SolverType >( Base::SolverType::Reservoir ) >( m_solvers ) = reservoirSolver;
+
+  CompositionalMultiphaseWell * wellSolver =
+    &this->getParent().template getGroup< CompositionalMultiphaseWell >( m_wellSolverName );
+  std::get< toUnderlying< typename Base::SolverType >( Base::SolverType::Well ) >( m_solvers ) = wellSolver;
+
+  Base::getWellSolver()->setFlowSolverName( m_reservoirSolverName );
+
+  integer const & useMassFlow = Base::getReservoirSolver()->template getReference< integer >( CompositionalMultiphaseBase::viewKeyStruct::useMassFlagString() );
+  integer const & useMassWell = Base::getWellSolver()->template getReference< integer >( CompositionalMultiphaseWell::viewKeyStruct::useMassFlagString() );
+  GEOSX_THROW_IF( useMassFlow != useMassWell,
+                  GEOSX_FMT( "CompositionalMultiphaseReservoir '{}': the input flag {} must be the same in the flow and well solvers, respectively '{}' and '{}'",
+                             this->getName(), CompositionalMultiphaseBase::viewKeyStruct::useMassFlagString(),
+                             Base::getReservoirSolver()->getName(), Base::getWellSolver()->getName() ),
+                  InputError );
```

Comment:
yummy 😋

Reply:
Now this is lighter, it looks like this:
https://github.com/GEOSX/GEOSX/blob/1dc1cf9ca37833654eefc891eaad228b6dd03dc1/src/coreComponents/physicsSolvers/multiphysics/CoupledReservoirAndWellsBase.cpp#L78-L83
for now